make featuresAt return symbols and use style properties

ansis opened the pull request:
This adds currently visible text and icons to featuresAt results.

Style properties are used to return features whose rendered representation matches the query.

query test pr: https://github.com/mapbox/mapbox-gl-test-suite/pull/77

fix #303
fix #316
fix #862
fix #902
fix #1822
fix #2100
fix #2103

This increases worker memory usage by around 10% (a couple of MB).
1/3 of the increase comes from retaining the CollisionTile.
2/3 of the increase comes from retaining VectorTileFeatures.


SymbolBucket merges lines to improve placement. This means that sometimes there is more than one VectorTileFeature responsible for that symbol. It currently returns the first VectorTileFeature and ignores the others that were merged into it. Should it do something else?

:eyes: @jfirebaugh

jfirebaugh commented:
Query tests? :wink: 

cc @mourner

ansis commented on js/source/worker.js:
```diff
@@ -153,7 +171,25 @@ util.extend(Worker.prototype, {
+            }
+
+            callback(null, tile.featureTree.query(params, this.styleLayersByID));
         } else {
```
@lucaswoj I just added these style layer creations and recalculations here as part of the style-property-aware featuresAt but I don't think they should stay here. Some of this is a duplication of what `Style` does but we don't need a full `Style` here. Any ideas on how this should be split out? Maybe a `StyleLayers`?

lucaswoj replied:
Can you read the `Style` from `tile.symbolBuckets[i].style`? We could even add a `symbolBucketsById` property, if you think that'd have better perf.

lucaswoj replied:
We currently create an instance of `Style` per bucket. That now strikes me as silly. We should create a single persistent `Style` per `Worker` and update it whenever we receive `set layers`

lucaswoj replied:
We should create a single persistent `Style` per `Worker` and update it whenever we receive `set layers`. We can add flags to disable any extra overhead on `Style` (i.e. fetching TileJSON).

averas commented:
Will this addition allow me to _distinguish_ "geographical" hits from text/icon hits? If you look at my example in #1882 where I have three icons based on a single geographical point; would I be able to sort out a click on the middle icon (where the point actually is) or will it return all three layers with no way to tell that it was one icon hit and three "geo" hits?

ansis commented:
> Will this addition allow me to distinguish "geographical" hits from text/icon hits?

After this and https://github.com/mapbox/mapbox-gl-js/issues/316 there will be no more "geo" hits if I'm understanding you correctly. All hits would be based on the rendered representation of features, not the source geometry. In your case, if you click in the center you would get a hit only for the center icon.

averas commented:
@ansis Exactly what I was hoping for! Thanks!

ansis commented:
@jfirebaugh added query tests: https://github.com/mapbox/mapbox-gl-test-suite/compare/query-tests?expand=1


This is ready for review

lucaswoj commented:
I'm generally feeling good about this code but I'm not sure we can afford to suddenly switch from "geographical" hits to "symbol" hits. Have you thought any more about making that a separate API @ansis?

lucaswoj commented:
Can we call the APIs `featuresAt` and `dataFeaturesAt` to emphasize their similarity while highlighting their differences? 

If I could do a clean-slate rename, I might put a verb in the name (and remove the preposition at the end of the name) like `findFeatures` and `findDataFeatures` but I think backwards compatibility is more important here.

ansis commented:
> Can we call the APIs featuresAt and dataFeaturesAt to emphasize their similarity while highlighting their differences?

The current implementation doesn't have very many similarities. They accept different parameters and return different things. I think the names should be very different.

> If I could do a clean-slate rename, I might put a verb in the name (and remove the preposition at the end of the name) like findFeatures and findDataFeatures but I think backwards compatibility is more important here.

I think it's ok to rename it since we're already breaking it's behavior heavily. I don't like `featuresAt` as a name either. I'm also wondering if `featuresAt` and `featuresIn` should be merged.

lucaswoj commented:
We need to take a moment to think about the linguistic semantics of these API methods

 - `featuresAt`
 - `featuresIn`
 - `getSourceTileData`

In particular, we need to

 - emphasize the ways these methods are similar (they introspect the map's features)
 - highlight the ways these methods are different (query geometry shape, feature type)
 - keep these method names as short as possible
 - follow conventions in API method naming
 - follow conventions in English grammar (ending a phrase with a preposition is bad practice)

If we can tolerate a clean-slate API redesign (preserving the existing methods as "deprecated" for a few releases) then I propose the following method names, each of which can handle both bbox and point-and-radius queries. 

 - `findStyledFeatures(options)`
 - `findFeatures(options)`

lucaswoj commented:
https://github.com/mapbox/mapbox-gl-js/issues/2125#issuecomment-184328553 makes me think we might want to drop `featuresAt` entirely

jfirebaugh commented:
For rendering queries, I think we'd keep `featuresAt` and `featuresIn` equivalents, but drop the `radius` option. The goal is that they both become exact queries of the rendered representation.

You could implement `featuresAt` in terms of `featuresIn` with a degenerate bounding box consisting of a single point, but I think we'd still want it for API clarity.

ansis commented:
> If we can tolerate a clean-slate API redesign (preserving the existing methods as "deprecated" for a few releases) then I propose the following method names

I'm fine with a breaking API redesign without deprecation. We don't want to support the old version for any meaningful length of time so we might as well drop it now.

>  each of which can handle both bbox and point-and-radius queries.

I think we should leave `getSourceTileData` (or whatever it's called) as a low level api that just returns all the data and doesn't do any geometric filtering (https://github.com/mapbox/mapbox-gl-js/commit/6c1b03cf9d1b043470e2c159f38a3c5635dcef5a):
- we aren't that clear on what the use cases for this could be so it's better to provide the low level access and add more once we learn about how it's used
- the source data is tiled which means there are a lot of gotcha's with clipped and duplicate features. Returning each tile as a complete and separate feature collection is upfront about this.

> findStyledFeatures(options)
> findFeatures(options)

I'm a bit afraid of overloading the name `Features` with multiple meanings, but if we need to use it we should at least make `findFeatures` into something more specific like `findSourceFeatures`.

>  but drop the radius option

Could the radius option still be useful for less precise selection like touch?

> You could implement featuresAt in terms of featuresIn with a degenerate bounding box consisting of a single point, but I think we'd still want it for API clarity.

Sounds good to me.

jfirebaugh commented:
> Could the radius option still be useful for less precise selection like touch?

For that use case you want to use `featuresIn` so that you don't run into #2125.

averas commented:
Perhaps there is some inspiration and experience that could be drawn from OpenLayers 3 here. They have been doing pixel perfect hit detection of vector data since their early releases:

> HIT DETECTION
The new vector implementation includes a technique for detecting features at a given pixel. This technique involves re-rendering features, one by one, in a 1 pixel by 1 pixel canvas, and, for each feature, testing the presence of a color using Canvas’ getImageData function. This technique is very robust (pixel-perfect), and it is proven to be very efficient as well.

ansis commented:
@averas thanks for the suggestion. We considered that approach but some experiments showed that rendering each feature one-by-one would be too slow and that reading back the pixels with `gl.readPixels` is slow and blocks the main thread . It could work well with a limited number of features but our goal right now is to make every map feature interactive. Calculating intersections on the cpu has other downsides but hopefully it will work out.

ansis commented:
I dropped `radius`, combined `featuresAt` and `featuresIn` internally, and fixed `featuresIn` when when the map is rotated or tilted. The main remaining things are:
- review
- decided on external api and naming

mourner commented:
@ansis reviewed the feature tree geometry code. It generally looks good, the only thing that caught my attention is that you rebuild full geometries on every query to account for offsets/etc., which is expensive. I'd prefer the intersection code to account for offsets during the checks, so that we only touch a fraction of all geometries due to early rejects/accepts.

ansis commented:
I'd like to get this merged this week. The main remaining thing is api/naming.

#### This branch currently has:
`map.featuresAt(point, { layer: 'name', $type: 'LineString', includeGeometry: true }, cb)`
`map.featuresIn(box, { layer: 'name, $type: 'LineString', includeGeometry: true }, cb)`
`map.getSourceTileData(sourceID, { 'source-layer': 'roads' }, cb)`


#### Questions:
Should we drop `$type`?
my answer: Yes, drop it. If we need filtering we should add full style-spec filters but let's not add that yet.

The internal implementation supports querying against lines and polygons, not just points and boxes. Should we expose that?
my answer: No, let's keep the external api simple. I'm not sure what the use cases are and we don't have well defined point geometry types yet.

#### Naming ideas

The equivalent dom api is [`elementFromPoint`](https://drafts.csswg.org/cssom-view/#dom-document-elementfrompoint) and `elementsFromPoint` with a tiny bit of `querySelector`.

https://github.com/mapbox/mapbox-gl-js/pull/2052#issuecomment-184805555:
> each of which can handle both bbox and point-and-radius queries.
> - findStyledFeatures(options)
> - findFeatures(options)

Other verbs: `query`, `search`.

#### Proposal

`map.featuresFromPoint(point, { layer: 'name', includeGeometry: true }, cb)`
`map.featuresFromBox(box, { layer: 'name', includeGeometry: true }, cb)`
`map.getSourceTileData(sourceID, { 'source-layer': 'roads' }, cb)`

and maybe add single feature versions: `map.featureFromPoint` and `map.featureFromBox`.

averas commented:
My (non-native) english would interpret <code>featuresFromPoint</code> as a distance type query...

lucaswoj commented:
I think your high level API design is sound. We should amend the proposal to

1. Make all method names into verb phrases (e.g. adding `find` or `search` to the names). This clearly marks these properties as methods and unambiguously states what the methods do. @averas has already noted the potential for "featuresFromPoint" to be misunderstood as a distance-measuring method. Though it is ubiquitous, I don't think the DOM is a good source of inspiration for API naming. 

2. Avoid using the verb "get" in the name of a relatively expensive API call. Convention dictates this is used for cheap getters / setters.

3. Seriously consider using the "features" terminology for all of these methods. We call them features internally everywhere, whether rendered or not. The vector tile spec calls them features, whether rendered or not. Our support team and users call them features, whether rendered or not. "data" strikes me as a much more vague word than "features" in the context of mapbox-gl-js. 

## Proposal 

 - `findFeaturesAtPoint(point, { layer: 'name', includeGeometry: true }, cb)`
 - `findFeaturesAtBox(box, { layer: 'name', includeGeometry: true }, cb)`
 - `findSourceFeatures(sourceID, { 'source-layer': 'roads' }, cb)`

lucaswoj commented:
Is it hard to combine `findFeaturesAtPoint` and `findFeaturesAtBox` into a single API that branches based on whether the first argument is a point or a box?

ansis commented:
> Make all method names into verb phrases (e.g. adding find or search to the names). This clearly marks these properties as methods and unambiguously states what the methods do. @averas has already noted the potential for "featuresFromPoint" to be misunderstood as a distance-measuring method. Though it is ubiquitous, I don't think the DOM is a good source of inspiration for API naming.

I completely agree

> Avoid using the verb "get" in the name of a relatively expensive API call. Convention dictates this is used for cheap getters / setters.

I agree. I don't think we should use "find" though because the method isn't finding or searching for anything. It's just returning everything that's loaded.

> Seriously consider using the "features" terminology for all of these methods. We call them features internally everywhere, whether rendered or not. The vector tile spec calls them features, whether rendered or not. Our support team and users call them features, whether rendered or not. "data" strikes me as a much more vague word than "features" in the context of mapbox-gl-js.

Yeah, "data" is too vague. I don't think we should use just "features" for both of them though. "features" means something different in each case so we should try to find different words.

> Is it hard to combine findFeaturesAtPoint and findFeaturesAtBox into a single API that branches based on whether the first argument is a point or a box?

Not technically, but https://github.com/mapbox/mapbox-gl-js/pull/2052#issuecomment-184833048 . I don't have an opinion either way.

ansis commented:
@averas thanks for the input

We decided on:

```
queryRenderedFeatures(pointOrBox, { layer: 'name', includeGeometry: true }, cb)
querySourceFeatures(sourceID, { 'source-layer': 'roads' }, cb)
```

1f6b464 changes the return format of querySourceFeatures
 7c0b790 replaces $type with filter

Everything should be implemented. @lucaswoj want to review?

jfirebaugh commented:
`source-layer` looks kind of awkward with `includeGeometry` being camel case. I understand the reluctance to overload `layer`, but... `querySourceFeatures(sourceID, { layer: 'roads' }, cb)`?

ansis commented:
> source-layer looks kind of awkward with includeGeometry being camel case. I understand the reluctance to overload layer, but... querySourceFeatures(sourceID, { layer: 'roads' }, cb)?

yeah, and to be consistent with the style spec. `sourceLayer` maybe? I'm fine with `layer` but prefer `sourceLayer`.

lucaswoj commented:
:+1: for `sourceLayer`

ansis commented:
remaining:
- [ ] update examples
- [ ] update changelog
- [ ] explore if this can be made fast enough and transferrable enough to be run synchronously on the main thread. This isn't a tiny amount of work but I want to look into this now because it would be a breaking change.

jfirebaugh commented:
> it would be a breaking change

You mean, if we removed callbacks from the interface? I don't think we should do that -- even if it can be made sync internally, I think we should preserve the option of making it async again in the future, and do ~`setTimeout(callback.bind(results), 0)`.

mourner commented:
@ansis yeah, I would also suggest to settle on worker-based featuresAt for now and revisit later while keeping compatibility. The PR is already pretty big.

lucaswoj commented:
fixes #1785

averas commented:
I though I'd raise an issue I've ran into a couple of times before this PR is merged, since a solution might also be breaking. When calling featureIn on multiple layers, such as <code>{layers: ['layer1', 'layer2']}</code> on a large data set (say a couple of hundred thousand points) there is a certain cost and time associated with such as call, which is natural.

In many of my use cases, however, I want to pass the results, or an aggregation of the results (for example a count), on to a visualization (say d3), but before I can do that I have to filter the data set again! Why? Because the result returned by featuresIn is flattened to a single array with the layer information present on each object, such as:

```
[
{layer: {...}, properties: {...} ... },
{layer: {...}, properties: {...} ... },
{layer: {...}, properties: {...} ... }
]
```
When passing this data on I first have to do a <code>features.filter(... layer.id === "...")</code>.

Would it make sense to evaluate whether it would actually be better to return a hierarchical representation of the results where features are nested under their respective layer?

ansis commented:
rebased on master. 8bdf3c7f0926a02b04064153a4685af2ab1ea4a5 was the latest before the rebase.
